File: frontend/src/types/template.ts
export interface TemplateSummary {
  name: string;
  category: string;
  description: string;
  version: string;
  field_count: number;
}

export interface ConstraintDef {
  min?: number | null;
  max?: number | null;
  min_age?: number | null;
  max_age?: number | null;
  values?: string | null;
  right_digits?: number | null;
  format?: string | null;
  start?: string | null;
  end?: string | null;
}

export interface FieldDef {
  name: string;
  type: string;
  generator: string;
  unique: boolean;
  formula?: string | null;
  constraint?: ConstraintDef | null;
}

export interface RelationshipDef {
  type: string;
  source: string;
  target?: string | null;
}

export interface TemplateMeta {
  description: string;
  version: string;
}

export interface Template {
  name: string;
  category: string;
  meta: TemplateMeta;
  fields: FieldDef[];
  relationships: RelationshipDef[];
}

File: backend/app/schemas/template.py
from pydantic import BaseModel, Field


class ConstraintDef(BaseModel):
    min: float | None = None
    max: float | None = None
    min_age: int | None = None
    max_age: int | None = None
    values: str | None = None
    weights: str | None = None
    right_digits: int | None = None
    format: str | None = None
    start: str | None = None
    end: str | None = None


class FieldDef(BaseModel):
    name: str
    type: str
    generator: str
    unique: bool = False
    formula: str | None = None
    null_probability: float | None = None
    constraint: ConstraintDef | None = None
    condition: str | None = None


class RelationshipDef(BaseModel):
    type: str
    source: str
    target: str | None = None


class TemplateMeta(BaseModel):
    description: str = ""
    version: str = "1.0"


class Template(BaseModel):
    name: str
    category: str = "General"
    meta: TemplateMeta = Field(default_factory=TemplateMeta)
    fields: list[FieldDef] = Field(default_factory=list)
    relationships: list[RelationshipDef] = Field(default_factory=list)


class TemplateSummary(BaseModel):
    name: str
    category: str
    description: str
    version: str
    field_count: int
